src/publishers/mock.rs

```diff
@@ -59,14 +59,26 @@ impl MockPublisher {
             uuid, topic
         );
     }
+
+    /// Get a view over the messages that have been published to this publisher
+    // assert_message_published is not flexible enough for testing (for example, checking that some
```

Comment:
I think its perfectly fine to expose a method to assert a message has not been published. Are there any other use-cases we need to have satisfied?

Reply:
that feels like such a constrained testing API though. returning something abstract like a slice or iterator allows all the flexibility of `find`/`filter`/`nth`/`eq`. How to express "message X was published before Y" for example? Yes, certain publishers don't ensure publish ordering, but a test can infer internal behavior when assuming a well-behaved publisher. I use this in one of my tests, to see that a retry message was published after an initial batch

Reply:
In that case, please do not rely on the fact that the messages in the mock publisher are stored sequentially (i.e. don't return a slice), even for an internal API. Having (even if internal) tests rely on the fact that messages are stored sequentially would make changes to the `MockPublisher` potentially difficult in the future.

I'm comfortable with having an API that returns a `impl Iterator<Item=(Topic, Message)>` or some such, though. That seems useful and harmless enough that I think it ought to be _public_, even.

Reply:
Yeah I wanted to return an `impl Iterator` too, but the lock made it weird and contorted. I'll try again and see what works

Reply:
alright, I made it return a type that can provide an iterator. That seems less bad all around